proving_system/stark/src/air/trace.rs

```diff
@@ -55,14 +58,70 @@ impl<F: IsTwoAdicField> TraceTable<F> {
             .collect()
     }
 
+    pub fn get(&self, step: usize, col: usize) -> FieldElement<F> {
+        let idx = step * self.n_cols + col;
+        self.table[idx].clone()
+    }
+
     pub fn compute_trace_polys(&self) -> Vec<Polynomial<FieldElement<F>>> {
         self.cols()
             .iter()
             .map(|col| Polynomial::interpolate_fft(col))
             .collect::<Result<Vec<Polynomial<FieldElement<F>>>, FFTError>>()
             .unwrap()
     }
+
+    /// Validates that the trace is valid with respect to the supplied AIR constraints
+    pub fn validate<A: AIR<Field = F>>(&self, air: &A) {
+        info!("Starting constraints validation over trace...");
+
+        // --------- VALIDATE BOUNDARY CONSTRAINTS ------------
+        air.boundary_constraints()
+            .constraints
+            .iter()
+            .for_each(|constraint| {
+                let col = constraint.col;
+                let step = constraint.step;
+                let boundary_value = constraint.value.clone();
+                let trace_value = self.get(step, col);
+
+                if boundary_value != trace_value {
+                    error!("Boundary constraint inconsistency - Expected value {:?} in step {} and column {}, found: {:?}", boundary_value, step, col, trace_value);
+                }
+            });
+
+        // --------- VALIDATE TRANSITION CONSTRAINTS -----------
+        let root_order = air.context().trace_length.trailing_zeros();
+        let _trace_primitive_root: FieldElement<F> =
+            F::get_primitive_root_of_unity(root_order as u64).unwrap();
```

Comment:
I think this instruction can be removed.

Reply:
good catch, thanks!